src/kernel.cr

```diff
@@ -1,18 +1,88 @@
 {% if flag?(:win32) %}
+  # The standard input file descriptor. Contains data piped to the program.
   STDIN  = IO::FileDescriptor.new(0)
+
+  # The standard output file descriptor.
+  #
+  # Typically used to output data and information.
+  #
+  # NOTE: Gets flushed when a newline is written.
   STDOUT = IO::FileDescriptor.new(1).tap { |f| f.flush_on_newline = true }
+
+  # The standard error file descriptor.
+  #
+  # Typically used to output error messages and diagnostics.
+  #
+  # NOTE: Gets flushed when a newline is written.
   STDERR = IO::FileDescriptor.new(2).tap { |f| f.flush_on_newline = true }
 {% else %}
   require "c/unistd"
 
+  # The standard input file descriptor. Contains data piped to the program.
   STDIN  = IO::FileDescriptor.from_stdio(0)
+
+  # The standard output file descriptor.
+  #
+  # Typically used to output data and information.
+  #
+  # NOTE: Gets flushed when a newline is written.
   STDOUT = IO::FileDescriptor.from_stdio(1).tap { |f| f.flush_on_newline = true }
+
+  # The standard error file descriptor.
+  #
+  # Typically used to output error messages and diagnostics.
+  #
+  # NOTE: Gets flushed when a newline is written.
   STDERR = IO::FileDescriptor.from_stdio(2).tap { |f| f.flush_on_newline = true }
 {% end %}
 
+# The name, the program was called with.
 PROGRAM_NAME = String.new(ARGV_UNSAFE.value)
-ARGV         = Array.new(ARGC_UNSAFE - 1) { |i| String.new(ARGV_UNSAFE[1 + i]) }
-ARGF         = IO::ARGF.new(ARGV, STDIN)
+
+# An array of arguments passed to the program.
+ARGV = Array.new(ARGC_UNSAFE - 1) { |i| String.new(ARGV_UNSAFE[1 + i]) }
+
+# An `IO` for reading files from `ARGV`.
```

Comment:
I would suggest to start this description with a usage example from an executable in a terminal.

Like
```
$ crystal build cli.cr
$ echo "foo" | ./cli
$ ./cli file.txt
```

What will `ARGF.gets_to_end` will have in both scenarios?
What will the behaviour be if multiple arguments are given or not?
What will happen if some arguments are not files and what can be done in that scenario.

I things that would be a more easy to understand flow for ARGF.

Reply:
Like this?
```cr
# An `IO` for reading files from `ARGV`.
#
# Usage example:
#
# The program: (`program.cr`):
# ```
# puts ARGF.gets_to_end
# ```
#
# A file to read from: (`file`)
# ```text
# 123
# ```
#
# ```
# $ crystal build program.cr
# $ echo "hello" | ./program
# hello
# $ ./program file
# 123
# $ ./program file file
# 123123
# $ ./program unknown
# Unhandled exception: Error opening file 'unknown' with mode 'r': No such file or directory (Errno)
# ...
# ```
#
# After a file from `ARGV` has been read, it's removed from `ARGV`.
#
# You can manipulate `ARGV` yourself ...
```